Initial Codebase

jaybosamiya opened the pull request:
This PR essentially adds all the initial code in (but was made as a PR to make it easier to comment and review).

Combined with another PR in `bap` and another PR in `opam-repository`, this will essentially make the BAP-IDA collaboration a lot more smooth and tries to ensure that it is kept quite robust to differing versions of IDA. Also, the plugin loader allows us to seamlessly have users get updated plugins whenever necessary (they just need to upgrade through pip), and also allows bap to access the latest IDA on the system without forcing a recompile.

BTW, after this code is installed, the user can actually use BAP through IDA with only minor additions to get IDAPython onto IDA Demo (from https://github.com/idapython/src/). This allows a much larger user base to be able to use BAP along with IDA (though it automatically disables the pseudocode view plugins, since decompiler doesn't exist on IDA Demo).

jaybosamiya committed:
PyPI prefers .rst

jaybosamiya committed:
Add the code in

The code was developed earlier on https://github.com/jaybosamiya/bap_ida_python
and then copied over, which is why the history is just one commit :D

ivg commented on README.rst:
```diff
@@ -0,0 +1,16 @@
+
+Now all that remains is to inform IDA about where the scripts are and how to access BAP. This process is simplified greatly using the plugin installer. The installer can be run using::
+
+    python -c 'from bap_ida_python.loader.installer import main; main()'
```
can we actually pack this into a python script?
Also, is it possible to do this in a process of `pip install` or it is an interactive script?

jaybosamiya replied:
It is packed as a python script (see [loader/installer.py](https://github.com/jaybosamiya/bap-ida-python/blob/1c407ee7a9d4f3390db8ca8f1a24ce86ec07e1d2/bap_ida_python/loader/installer.py)), but since it is run only once, I did not think it made sense to have it install a new command itself onto the shell PATH.

Also, I initially packed it in as a post-install command. However, usually the `pip` command is run via `sudo` and in such a scenario, `bap` is not really on the `$PATH` if installed through `opam`.

This script is interactive only of the person does not have the `which` command on their system. Otherwise, it automatically generates all necessary information.

ivg replied:
I feel that `python -c 'from bap_ida_python.loader.installer import main; main()'` magic is not user friendly.

jaybosamiya replied:
It can be simplified down further if necessary. I spent quite some time to see if we could automate it through pip as a post-install, but believe me, this turns out to be more robust on more systems. By tagging it along as a post-install, it fails under certain scenarios, and works fine under others. Hence, I prefer the more robust (though more verbose) solution. Also, it is a single copy-paste.

Alternative is to install a `bap-ida-python-plugin-installer` or similarly named utility which I find to be overkill, since it is a one-shot thing.

ivg replied:
well, if you don't want to clobber the path, then at least `python -m bap_ida_python.install` would be much easier to type and to understand.

jaybosamiya replied:
Ah, well this works: `python -m bap_ida_python.loader.installer`

I can update the README with this.

ivg commented on bap_ida_python/loader/installer.py:
```diff
@@ -0,0 +1,88 @@
+        return None
+
+
+def install_bap_config():
```
can we do this without a config file at this stage?

Configuration is a complex feature, that should be designed carefully, not just added on per need basis. We should think about the general approach for storing static configuration data, that will work uniformly across all plugins and parts of the platform. 

Also, it looks really weird, that some non-essential component, is responsible for installation of BAP config. The config should be installed by the BAP, and owned by it. 

Finally, I feel really bad about establishing a communication channel between two software components via a configuration file. A program shouldn't modify the config file, it is the realm of a human. It can be generated by a program, but then left to a human. Even worser, if one program is changing a configuration file of another program.

jaybosamiya replied:
Due to the non-standard path of `bap` and non-standard path of `ida`, it is going to be necessary to use something to communicate what is where.

I have not added it on a per-need basis, and have spent quite some time thinking about how the config would be work. And that is why I have chosen the format `(BAP (k1 v1) (k2 v2) (k3 v3) ...)` (i.e. an S-Expression starting with `BAP`, which stores key-value pairs) as the format. Please suggest alternatives if you find necessary.

Also, any suggestions for how to pass the info better would definitely be appreciated :)

One minor modification which might make more sense then, would be to use `.bap/ida-python-config` for all ida-python related stuff. In that scenario, we would need to modify [this line](https://github.com/BinaryAnalysisPlatform/bap/pull/467/files#diff-38344486842e6d7253058269badbb7bbR5) and [this line](https://github.com/BinaryAnalysisPlatform/bap-ida-python/pull/1/files#diff-ec5e6bc658f71f9c3c08912267a6d45eR10).

ivg replied:
Well, of course, I will provide suggestions and we will workout the design together. I'm just highlighting the obvious violations of a common system design principles. 
1. The configuration file is not an IPC media.
2. Configurations files shouldn't be changed by 3rd party software
3. The configuration system should be designed as an integral part of a platform, not evolve based on our needs.
4. The interpackage state information should be stored in the OS/package manager layer. 

So, there're lot of things to be discussed here.

jaybosamiya replied:
Once I read your points, I realized that we are in complete agreement, and that the current configuration system has issues. However, my suggestion to place it in `.bap/ida-python-config` is still valid due to the following (numbered wrt your points):
1. This is not being used as an IPC media. It is only to store the configuration. "Where is something found" is a configuration. That is all we are storing.
2. Agreed. However, since both bap and bap-ida-python are us (i.e. BAP org) itself, it isn't 3rd party. Hence using a common `.bap/` makes sense. However, since it is not the main program, we can have it in a `ida-python-config` file. Also, we could have it generate it if it doesn't exist, but not modify it if it does. In such a scenario, it'd make sense, right?
3. The information of where IDA is and where BAP is, is an integral part of the `bap-ida-python` system. Without knowing that, it would be impossible to make any calls. Hence, my proposal for the `.bap/ida-python-config`.
4. When multiple package managers are used (`opam`/source for BAP, Installer/manual for IDA Pro, `pip` for bap-ida-python), there is no package manager layer. Hence, we are left with either the OS doing some black magic for us, or us storing things in files, and this is why I chose to store it in files.

Yep, agreed, we need to discuss quite a bit, and I think it might be better if we take this offline :)

dbrumley replied:
We keep hitting this roadblock.  Recompiling every time IDA path changes in BAP is unacceptable to me.  Jay is hitting roadblocks trying to support IDA, which is a killer application.

These two problems must be addressed; practical usefulness is  more important than philosophy.

What suggest do you have?  Why does creating a configuration file require intense thought at this point, as opposed to using something simple and iterating as we figure out how we use?

ivg replied:
There is no roadblock here to hit, and the question is not about philosophy. Be assured, we with Jay will be able to find out a solution, that will satisfy your needs. Configuration file doesn't contradict with my philosophy, if this is what you thought. This is merely technical issues, and all of them has straightforward solution.

jaybosamiya committed:
Simplify installation

ivg commented on bap_ida_python/loader/installer.py:
```diff
@@ -0,0 +1,88 @@
+
+    print "[+] Found bap installed at", bap_path
+
+    bap.config.set('bap_executable_path', bap_path)
```
Am I reading this correct, that you're storing path to bap in bap's config?

jaybosamiya replied:
Please read my above comment about modifying the config to be made `.bap/ida-python-config`.

ivg replied:
I read, and either I didn't understand it, or it doesn't answer my question.

jaybosamiya committed:
Stop using PIP + Use IDADIR/cfg for config + Use IDA style config file

+ Moves away from the `pip` based install (which might cause issues due to differing runtimes of IDA (x86) and the system (x86-64); they currently don't cause problems, but if some dependency later on has some C binary requirement, then it might cause issues).
+ Restructures the `bap_ida_python` module into a module called just `bap`
+ Removes some unnecessary global config
+ Moves config from global path (on `$HOME/.bap/config`) to local IDA config (`IDADIR/cfg/bap.cfg`)
+ Uses the standard config format that other files in `IDADIR/cfg` use
+ Automatically asks the user after installation to provide BAP path (and stores it in the config)
    + There is some "automagic" script that I've added that also provides a helpful default to the user if IDA can automatically detect where bap is. Even then, it gives the user a chance to chose the path.

ivg commented on plugins/bap/utils/config.py:
```diff
@@ -0,0 +1,66 @@
+                return line.split()[1]
+    return default
+
+def set(key, value, section='default'):
```
Do we really need `set` function?

jaybosamiya replied:
Yes, we need this since we have the functionality for the user to have a prompt on the first usage of the plugin, and this will set the configuration file. Since we have a copy-paste based system, we either need to do it like this, or give the user instructions saying "create a file in .... with the bap path at ..." etc. which I do not think is good user experience.

ivg commented on plugins/bap/utils/run.py:
```diff
@@ -0,0 +1,127 @@
+        return askyn_c(ASKBTN_CANCEL, msg) == ASKBTN_YES
+
+    while True:
+        bap_path = idaapi.askstr(0, default_bap_path, 'Path to bap')
```
maybe there is some file dialog in IDA (that is portable across 6.6-6.9)?

ivg commented:
Looks good to me, green light.

jaybosamiya committed:
Use file dialog for getting bap path

jaybosamiya commented:
Added a file dialog to get bap path for initial configuration. Merging.
